cpp/grammar_compiler.cc

```diff
@@ -335,90 +368,77 @@ AdaptiveTokenMask GrammarMatcherForTokenMaskCache::GetAdaptiveTokenMask(
-        compile_structural_tag_cache_(GetCompileStructuralTagCacheFunc(cache_enabled_)),
-        compile_regex_cache_(GetCompileRegexCacheFunc(cache_enabled_)),
-        compile_grammar_cache_(GetCompileGrammarCacheFunc(cache_enabled_)) {}
-
-  CompiledGrammar CompileBuiltinJSONGrammar();
-
-  CompiledGrammar CompileJSONSchema(
-      const std::string& schema,
-      bool any_whitespace,
-      std::optional<int> indent,
-      std::optional<std::pair<std::string, std::string>> separators,
-      bool strict_mode = true
-  );
-
-  CompiledGrammar CompileStructuralTag(
-      const std::vector<StructuralTagItem>& tags, const std::vector<std::string>& triggers
-  );
+namespace {
 
-  CompiledGrammar CompileRegex(const std::string& regex);
+using SchemaKey =
+    std::tuple<std::string, bool, std::optional<int>, std::pair<std::string, std::string>, bool>;
+using StructuralTagKey = std::tuple<std::vector<StructuralTagItem>, std::vector<std::string>>;
+using GrammarKey = std::pair<std::string, std::string>;
 
-  CompiledGrammar CompileGrammar(const Grammar& grammar);
+class GrammarCompilerBase {
+ public:
+  GrammarCompilerBase(const TokenizerInfo& tokenizer_info, int max_threads, bool cache_enabled)
+      : tokenizer_info_(tokenizer_info), max_threads_(max_threads), cache_enabled_(cache_enabled) {}
 
-  void ClearCache();
+  GrammarCompilerBase(const GrammarCompilerBase&) = delete;
+  GrammarCompilerBase(GrammarCompilerBase&&) = delete;
+  GrammarCompilerBase& operator=(const GrammarCompilerBase&) = delete;
+  GrammarCompilerBase& operator=(GrammarCompilerBase&&) = delete;
 
- private:
   /*!
    * \brief Build the tag dispatch fsm for the root rule and store in the compiled grammar.
    */
   void BuildTagDispatchFSM(Grammar grammar, const Grammar::Impl::RuleExpr& root_rule_expr);
 
   /*! \brief Multi-thread compile the grammar. */
-  CompiledGrammar MultiThreadCompileGrammar(Grammar grammar);
+  auto MultiThreadCompileGrammar(Grammar grammar) -> CompiledGrammar;
 
-  /*! \brief The cache for the compiled grammar of a JSON schema. */
-  using SchemaKey =
-      std::tuple<std::string, bool, std::optional<int>, std::pair<std::string, std::string>, bool>;
-  /*! \brief The cache function for the compiled grammar of a JSON schema. */
-  std::function<CompiledGrammar(const SchemaKey&)> GetCompileJSONSchemaCacheFunc(bool cache_enabled
-  );
-
-  /*! \brief The cache function for the compiled grammar for pure JSON. */
-  std::function<CompiledGrammar()> GetCompileBuiltinJSONGrammarCacheFunc(bool cache_enabled);
-
-  using StructuralTagKey = std::tuple<std::vector<StructuralTagItem>, std::vector<std::string>>;
-  /*! \brief The cache function for the compiled grammar for a given structural tag. */
-  std::function<CompiledGrammar(const StructuralTagKey&)> GetCompileStructuralTagCacheFunc(
-      bool cache_enabled
-  );
+  auto CompileJson() -> CompiledGrammar;
```

Comment:
```suggestion
  CompiledGrammar CompileJson();
```
ditto for other places